pkg/vpc/cleanup.go

```diff
@@ -30,49 +33,50 @@ func findAvailableNetworkInterfaces(provider api.ClusterProvider, spec *api.Clus
 			},
 		},
 	}
-	output, err := provider.EC2().DescribeNetworkInterfaces(input)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
-	}
+
 	securityGroupRE, err := regexp.Compile(fmtSecurityGroupNameRegexForCluster(spec.Metadata.Name))
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
+		return errors.Wrap(err, "failed to create security group regex")
 	}
-	networkInterfaces := []string{}
-	for _, eni := range output.NetworkInterfaces {
-		id := *eni.NetworkInterfaceId
-		for _, sg := range eni.Groups {
-			if securityGroupRE.MatchString(*sg.GroupName) {
-				logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
-				networkInterfaces = append(networkInterfaces, id)
-				break
-			} else {
+
+	var lastErr error
+
+	err = ec2API.DescribeNetworkInterfacesPagesWithContext(ctx, input, func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool {
```

Comment:
Yes, attempting deletions on the spot has proven to be messy in general. We might want to go all the way and create a list of deletion tasks like we do for most other things.

Reply:
So then, having this return an array can be a good API for now and for later if we decide to do that.

Reply:
The default page size I believe is around 1000. The use of pagination methods here is to handle the edge case where more than 1000 items exist. If we construct an array of matching items from all pages, all of them will be held in memory (granted, after filtering, they'd be a lot less).
The motivation is to avoid having to create a slice of matching ENIs and do the deletion in a single pass through the loop, instead of running another loop for deleting the items.

So for the normal case where less than 1000 items exist, this code is somewhat identical to the original code with the difference that it deletes ENIs as it finds them rather than creating a slice of matching items and running another loop.

>It also makes error handling and testing easier.

I agree that the code for testing this would be a bit different than the other test cases.

Reply:
I see your point that this is more efficient. I typically prefer readability and simplicity over performance, unless this is a very slow operation. Also, we don't have a garbage collection problem so I don't think creating new objects/arrays should be an issue.

I really value code maintenance, specially since we have more people working on this project and external contributors, so simplicity means less bugs, and faster ramp up for new joiners.

I won't block this as it is but I would ask you to consider this option for the reasons I mentioned :)

Reply:
I'm with Martina ok this one. Readability and maintenance over performance unless there is empirical evidence of performance being an actual problem.

Reply:
@martina-if I agree with your points. I'll make the proposed changes.